src/ratbag_emu/server/device_controller.py

```diff
@@ -18,20 +18,18 @@
 Devices
 '''
 def list_devices():
-    devices = []
+    device_list = []
 
-    for device_id, device in DeviceHandler.devices.items():
+    DeviceHandler.lock.acquire()
+    devices = DeviceHandler.devices.copy()
+    DeviceHandler.lock.release()
+    for device_id, device in devices.items():
         if not device:
             continue
 
-        devices.append({
-            'id':           device_id,
-            'name':         device.name,
-            'shortname':    device.shortname,
-            'input_nodes':  device.device_nodes
-        })
+        device_list.append(DeviceHandler.get_device(device_id))
```

Comment:
you can use list comprehension here

`device_list = [DeviceHandler.get_device(id) for id, device in devices.items() if device is not None]`

though I'm not sure it's that much of an improvement ;)